### ISSUE TITLE
Problems with test datasets in the SfM pipeline

### ISSUE THREAD

**Issue by saihv** [openMVG/openMVG#641]
I've been trying to use the SfM pipeline and the localization module for testing some of my datasets. My routine involves getting two views of the scene from two positions horizontally spaced apart, using them to reconstruct the environment into an sfm_data.bin file, and then localizing more images using this environment. 

OpenMVG modules run in order:
1. InitSfM_ImageListing -k <my_K_matrix>
2. ComputeFeatures
3. ComputeMatches -n ANNL2
4. IncrementalSfM
5. Sfm_Localization -q <queryImage>

One dataset (let's say dataset1) I tried with worked very well, and I was able to move the camera around and get accurate localization. But a few other datasets I tried with gave me multiple problems. I was able to fix the issues with some workarounds such as changing the feature detection method, using bruteforce instead of ANN etc., but the end localization result had a lot of error: so I am assuming the problem could be with my scene. Some of the problems I faced were:

a) After running ComputeMatches, no connecting line between the two views in the geometric_matches svg file. 
b) IncrementalSfM failing with either "zero tracks" error, or "MSE residual error: -1" 
c) SfM_Localization reporting "Cannot locate image. Points used by Robust Resection: (some small number, around 10-12)". (workaround: using the ULTRA preset in feature detection)

In dataset1, there were not a lot of points in the environment point cloud yet localization worked very well, but in dataset2, if I run feature detection with the ULTRA preset, I get a lot of points in the scene yet the end result has a lot of error. Sample results shown below, where the localization is performed as the camera moves forward and backward.

Good result with dataset1: (few points, accurate localization)
![2](https://cloud.githubusercontent.com/assets/2274262/19059379/536be94c-8994-11e6-9dde-842d1ca3652f.png)

Bad result with dataset2: (Lots of points, yet localization drifts)
![5](https://cloud.githubusercontent.com/assets/2274262/19059397/80c45df2-8994-11e6-9f87-bba104c72505.png)

I made sure that all the images captured were looking at the same scene as the initial views, so I'm somewhat confused as to what the issue could be. Any suggestions would be very helpful, thank you.

**Comment by pmoulon** [comment]
hi @saihv 

What is the difference between dataset1 and dataset2:
- are you using the same camera setup
- does it is the same building or another one?


This is strange that you do not have corresponding point between the adjacent view in your second dataset. Do you have other dataset to play with?

**Comment by saihv** [comment]
The cameras are the same, and the building is the same. Here are the two views I use to initialize from

Dataset 1

![1](https://cloud.githubusercontent.com/assets/2274262/19065503/cd262b24-89c8-11e6-8eee-85c6567b31de.png)

Dataset 2

![3](https://cloud.githubusercontent.com/assets/2274262/19065507/d109e654-89c8-11e6-8462-5a4a1ebd5a0a.png)

The one obvious difference is that in the second dataset the images look overexposed, but I am not sure if this can cause an issue, (especially because when I used the ULTRA preset, the feature matching works and you can see that dense point cloud), but at the end, the localization is inaccurate.

I am going to try with more datasets soon.

**Comment by saihv** [comment]
Just tried with a couple more kinds of scenes and I have similar issues. For instance, both these image pairs give me no geometric matches (there are putative matches but no geometric matches)

![screenshot from 2016-10-04 14 16 11](https://cloud.githubusercontent.com/assets/2274262/19092788/317dbab4-8a3d-11e6-9dda-95f0b3092058.png)

![screenshot from 2016-10-04 14 16 50](https://cloud.githubusercontent.com/assets/2274262/19092792/337e30aa-8a3d-11e6-9d9d-198649dbd596.png)

So basically I've been unable to replicate the success of the first dataset. Could the geometric matching fail because the 'baseline' between the views is not giving enough disparity (given how far the building is from the cameras, for example) so as to fit a model?

**Comment by pmoulon** [comment]
Can you share some image pairs in order to let the community make some experiments?

**Comment by saihv** [comment]
Attaching a zip with 3 pairs that gave me problems. The K matrix if necessary is "1332;0;626;0;1332;468;0;0;1"

[imagePairs.zip](https://github.com/openMVG/openMVG/files/512463/imagePairs.zip)

**Comment by saihv** [comment]
Here's another pair that gives me this problem when running IncrementalSfM.

Bundle Adjustment statistics (approximated RMSE):
 #views: 2
 #poses: 2
 #intrinsics: 1
 #tracks: 244
 #residuals: 976
 Initial RMSE: 1.23338
 Final RMSE: 0.625061
 Time (s): 0.0347211


=========================
 MSE Residual InitialPair Inlier: -1
=========================

[baseline.zip](https://github.com/openMVG/openMVG/files/515087/baseline.zip)

**Comment by pmoulon** [comment]
Will try to make some experiments next week

**Comment by saihv** [comment]
Thanks!

**Comment by pmoulon** [comment]
The five point estimation algorithm is used in OpenMVG incremental SfM pipeline or if you are using the essential matrix filter at main_ComputeMatches (-g e).

The coplanar degenerency can be a problem in your case (so perhaps Homography + relative pose estimation can help)

**Comment by saihv** [comment]
Interesting: for the latest pair I uploaded, (imagePair.zip) none of the models give me geometric matches (options -g f/e/h). The two nodes in geometric_matches.svg are always disconnected.

**Comment by pmoulon** [comment]
1. In order to try to have more points per images you can use the preset option -p (NORMAL/HIGH/ULTRA) in ComputeFeatures (use the -f 1 option if you want to rewrite old computed features).

2. You can simply rename or make a symbolic link from  matches.e.bin to matches.f.bin and launch IncrementalSfM (so you do not modify any code).

**Comment by saihv** [comment]
Hello, sorry to bother you again! I tried with HIGH and ULTRA presets on SIFT as well as AKAZE, but the matching never works (also tried using matches.e.bin with incrementalSfM, but because there are no matches anyway, neither global nor incremental SfM work). Tried it for another scene (indoors), and have the same problem. Not sure what's causing these serious problems with matching.

**Comment by pmoulon** [comment]
Did you try to run some tracking ...
In the VO folder you can try to track point in a serie of image: https://www.youtube.com/watch?v=EVq0ospyfW0

Perhaps it will help since the global motion is small and use local neighborhood can help.

**Comment by naveenjafer** [comment]
@pmoulon When does the error "  MSE Residual InitialPair Inlier: -1 " generally come up? I am trying to understand the behaviour or openMVG when just 2 images are fed to the system for similar localization uses that @saihv is using it for. No matter what feature descriptors I run it with(level and type) incremental SFM generally fails with " MSE Residual InitialPair Inlier: -1 ". I have tried images with wide enough baselines and also images with very short baselines, the issue is still the same. I have upto 23579 matches between my two image pairs that are selected for Initial Pair.

**Comment by pmoulon** [comment]
The -1 value comes from the fact that not valid relative pose have been found and so a default value is displayed.

In your case I would invite you to do the following:
1- Check that you scene have valid camera intrinsics, see the intrinsics field in the sfm_data.json
2- if you have valid intrinsic, understand why the robust relative pose is failing
  - angular filtering that remove all the corresponding points...

If your angle is too short the pair will be rejected since it is considered as unstable.

**Comment by naveenjafer** [comment]
@pmoulon The intrinsics is correct and I verified the file once more, because reconstruction works very well with all the images in the dataset, I am trying to sample a few at a time and see if ill be able to do reconstruction with just 2 of them, I have tried many different baseline distance/angle between images, but was unfortunately unable to come up with a reconstruction just using 2 images, After runnning various datasets it is an average of 5 images minimum generally that are needed for reconstruction.

**Comment by pmoulon** [comment]
Variability of the result is due to:
1. some randomness in the process (random number generator).
  - you can fix that by using a fix seed. I plan to work on that soon.
2. Threading -> openMP scheduling for some loop does not guarantee that operations are done in the same order.
  - here we have to see how OpenMP scheduling is set for every parallel for

Regarding your failure for 2 image. I continue to think that your pair is rejected due to a baseline issue. You can try to put some cout information there in order to understand.
https://github.com/openMVG/openMVG/blob/master/src/openMVG/sfm/pipelines/sequential/sequential_SfM.cpp#L480

**Comment by naveenjafer** [comment]
I'm not exactly allowed to upload anything from the restricted conditions I work in, it would be great if you could send me two images for which it works for you, I'll be able diagnose what exactly is going wrong with my baseline.

**Comment by pmoulon** [comment]
It should work with image pair from this dataset
https://github.com/openMVG/ImageDataset_SceauxCastle

But I recommend you to run in debug mode and to look step by step in order to understand why it fails. Since it works with more image it means that the initial pair is computed when you provide more image...

**Comment by naveenjafer** [comment]
@pmoulon The Issue was due to running it with multiple threads as you had suggested earlier "
    Threading -> openMP scheduling for some loop does not guarantee that operations are done in the same order. here we have to see how OpenMP scheduling is set for every parallel for" 
Running most of my images with single thread has solved the issue for me, I am curious as to why this is happening.

**Comment by pmoulon** [comment]
You can try to change some openMP schedule from dynamic to standard.

**Comment by naveenjafer** [comment]
@pmoulon images 100_7104.JPG and 100_7015.JPG are two good examples from the SceauxCastle dataset that fail when run with multiple threads and work fine when run with single thread. Parameters given were as follows. SIFT - HIGH - Incremental. In fact almost all the pairs in this dataset and others exhibit similar behaviour under multiple threads.
